tests/network/main.rs

```diff
@@ -0,0 +1,321 @@
+    repo_handle_2.stop().unwrap();
+
+    assert!(left_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(right_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(new_left_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(new_right_closed.load(std::sync::atomic::Ordering::Acquire));
+}
+
+#[test(tokio::test)]
+async fn test_streams_chained_on_replacement() {
+    // Create two repos.
+    let repo_1 = Repo::new(Some("repo1".to_string()), Box::new(SimpleStorage));
+    let repo_2 = Repo::new(Some("repo2".to_string()), Box::new(SimpleStorage));
+
+    // Run the repos in the background.
+    let repo_handle_1 = repo_1.run();
+    let repo_handle_2 = repo_2.run();
+
+    // Create a document for one repo.
+    let document_handle_1 = repo_handle_1.new_document();
+
+    // Edit the document.
+    document_handle_1.with_doc_mut(|doc| {
+        let mut tx = doc.transaction();
+        tx.put(
+            automerge::ROOT,
+            "repo_id",
+            format!("{}", repo_handle_1.get_repo_id()),
+        )
+        .expect("Failed to change the document.");
+        tx.commit();
+    });
+
+    let TinCan {
+        send: left_sink,
+        recv: left_stream,
+        sink_closed: old_left_closed,
+    } = tincan_to_nowhere();
+    repo_handle_1.new_remote_repo(repo_handle_2.get_repo_id().clone(), left_stream, left_sink);
+
+    let TinCan {
+        send: right_sink,
+        recv: right_stream,
+        sink_closed: old_right_closed,
+    } = tincan_to_nowhere();
+    repo_handle_2.new_remote_repo(
+        repo_handle_1.get_repo_id().clone(),
+        right_stream,
+        right_sink,
+    );
+
+    // Request the document.
+    let doc_handle_future =
+        tokio::spawn(repo_handle_2.request_document(document_handle_1.document_id()));
+
+    // Replace the peers.
+    let TinCans {
+        left:
+            TinCan {
+                send: left_sink,
+                recv: left_stream,
+                sink_closed: new_left_closed,
+            },
+        right:
+            TinCan {
+                send: right_sink,
+                recv: right_stream,
+                sink_closed: new_right_closed,
+            },
+    } = tincans();
+    repo_handle_1.new_remote_repo(repo_handle_2.get_repo_id().clone(), left_stream, left_sink);
+    repo_handle_2.new_remote_repo(
+        repo_handle_1.get_repo_id().clone(),
+        right_stream,
+        right_sink,
+    );
+
+    let doc_id = doc_handle_future.await.unwrap().unwrap().document_id();
+    assert_eq!(doc_id, document_handle_1.document_id());
+
+    // Stop the repos.
+    repo_handle_1.stop().unwrap();
+    repo_handle_2.stop().unwrap();
+
+    assert!(old_left_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(old_right_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(new_left_closed.load(std::sync::atomic::Ordering::Acquire));
+    assert!(new_right_closed.load(std::sync::atomic::Ordering::Acquire));
```

Comment:
Move these assertions to the drop impl?

Reply:
I agree that would be much nicer. Unfortunately it produces horrible error messages which halt the entire test suite because the panic in the drop is not unwindable. If you know a way around that then I think we should do the drop thing, but otherwise this seems like the least bad thing.

Reply:
Ok, agreed to keep it like this.